File: src/my_types.rs
use std::collections::{BTreeMap, HashSet, HashMap};
use rocket::serde;
use rocket::serde::{Deserialize, Serialize, de};

use crate::systemd_calendar_events;

#[derive(Deserialize)]
pub struct CasConfig {
    pub prefix_url: String,
}
fn default_separator() -> String { ".".to_owned() }
fn default_root_id() -> String { "".to_owned() }

fn remotes_periodicity_checker<'de, D>(d: D) -> Result<BTreeMap<String, RemoteConfig>, D::Error>
    where D: de::Deserializer<'de>
{
    let remotes : BTreeMap<String, RemoteConfig> = BTreeMap::deserialize(d)?;
    
    if systemd_calendar_events::next_elapses(
            remotes.values().map(|cfg| &cfg.periodicity).collect()
    ).is_err() {
        // there has been an error, retry one by one to know which one failed
        for (remote_id, cfg) in &remotes {
            if systemd_calendar_events::next_elapse(&cfg.periodicity).is_err() {
                let msg = format!("a valid periodicity for remote {}. Hint: validate it with ''systemd-analyze calendar ....''", remote_id);
                return Err(de::Error::invalid_value(de::Unexpected::Str(&cfg.periodicity), &msg.as_str()));
            }
        }
    }
    Ok(remotes)
}

fn ldap_config_checker<'de, D>(d: D) -> Result<LdapConfig, D::Error>
    where D: de::Deserializer<'de>
{
    let cfg : LdapConfig = LdapConfig::deserialize(d)?;
    
    if cfg.sgroup_sscfg_raw().is_none() {
        let msg = "''ldap.groups_dn'' to be listed in ''ldap.subject_sources''".to_owned();
        return Err(de::Error::invalid_value(de::Unexpected::Str(&cfg.groups_dn.0), &msg.as_str()));
    }
    Ok(cfg)
}

#[derive(Deserialize)]
pub struct StemConfig {
    pub filter: String,
    #[serde(default = "default_separator")]
    pub separator: String,
    #[serde(default = "default_root_id")]
    pub root_id: String,
}

#[derive(Deserialize, Serialize, PartialEq, Eq, Hash)]
pub struct SubjectSourceFlattenConfig {
    pub attr: String,
}

#[derive(Deserialize, Serialize, PartialEq, Eq, Hash)]
pub struct SubjectSourceConfig {
    pub dn : Dn,
    pub name : String,
    #[serde(skip_serializing_if="Option::is_none")]
    pub vue_template : Option<String>,
    #[serde(skip_serializing_if="Option::is_none")]
    pub vue_template_if_ambiguous : Option<String>,    
    pub display_attrs : Vec<String>,
    pub id_attrs : Option<Vec<String>>,

    #[serde(skip_serializing)]
    pub search_filter : String,
}

#[derive(Deserialize)]
pub struct AttrTexts {
    pub label: String,
    pub description: String,
}

#[derive(Deserialize)]
pub struct LdapConfig {
    pub url: String,
    pub bind_dn: Dn,
    pub bind_password: String,
    pub base_dn: Dn,
    pub groups_dn: Dn,
    pub stem_object_classes: HashSet<String>,
    pub group_object_classes: HashSet<String>,
    pub sgroup_filter: Option<String>, // needed if groupad does not own all groups in groups_dn
    pub stem: StemConfig,
    pub subject_sources: Vec<SubjectSourceConfig>,
    pub groups_flattened_attr: BTreeMap<Mright, String>,
    pub sgroup_attrs: BTreeMap<String, AttrTexts>,
}
#[derive(Serialize)]
pub struct LdapConfigOut<'a> {
    pub groups_dn: &'a Dn,
    pub subject_sources: &'a Vec<SubjectSourceConfig>,
}

impl LdapConfig {
    pub fn sgroup_sscfg_raw(&self) -> Option<&SubjectSourceConfig> {
        self.subject_sources.iter().find(|sscfg| sscfg.dn == self.groups_dn)
    }
    pub fn sgroup_sscfg(&self) -> &SubjectSourceConfig {
        self.sgroup_sscfg_raw().expect("internal error (should be checked as startup)")
    }

    pub fn to_js_ui(&self) -> LdapConfigOut {
        LdapConfigOut { groups_dn: &self.groups_dn, subject_sources: &self.subject_sources }
    }
}

#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(rename_all = "lowercase")]
pub enum RemoteDriver {
    #[cfg(feature = "mysql")]
    Mysql,
    #[cfg(feature = "oracle")]
    Oracle,
}

#[derive(Deserialize, Serialize, Clone)]
pub struct RemoteConfig {
    pub host: String,
    #[serde(skip_serializing_if="Option::is_none")]
    pub port: Option<u16>,
    pub driver: RemoteDriver,

    pub db_name: String,

    #[serde(skip_serializing)]
    pub user: String,
    #[serde(skip_serializing)]
    pub password: String,
    
    pub periodicity: String, // NB: checked by "remotes_periodicity_checker" below
}
#[derive(Deserialize)]
pub struct Config {
    pub trusted_auth_bearer: Option<String>,
    pub log_dir: Option<String>,
    pub cas: CasConfig,
    #[serde(deserialize_with = "ldap_config_checker")] 
    pub ldap: LdapConfig,
    #[serde(deserialize_with = "remotes_periodicity_checker")] 
    pub remotes: BTreeMap<String, RemoteConfig>,
}

#[derive(Deserialize, Serialize, PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Debug)]
#[serde(rename_all = "lowercase")]
pub enum Mright { Member, Reader, Updater, Admin }

#[derive(Deserialize, Serialize, PartialEq, Eq, PartialOrd, Ord, Debug, Copy, Clone)]
#[serde(rename_all = "lowercase")]
pub enum Right { Reader, Updater, Admin }

#[derive(Deserialize, Serialize, PartialEq, Eq, PartialOrd, Ord, Copy, Clone, Debug)]
#[serde(rename_all = "lowercase")]
pub enum MyMod { Add, Delete, Replace }

#[derive(Deserialize, Serialize, PartialEq, Eq, Debug, Hash, Clone, PartialOrd, Ord)]
#[serde(transparent)] 
pub struct Dn(pub String);

pub type MyMods = BTreeMap<Mright, BTreeMap<MyMod, DnsOpts>>;

impl From<String> for Dn {
    fn from(dn: String) -> Self {
        Dn(dn)
    }
}
impl From<&str> for Dn {
    fn from(dn: &str) -> Self {
        Dn(dn.to_owned())
    }
}

pub type MonoAttrs = BTreeMap<String, String>;

pub type SgroupsWithAttrs = BTreeMap<String, MonoAttrs>;

#[derive(Deserialize, Serialize, PartialEq, Eq, Debug, Clone, Default)]
pub struct DirectOptions {
    #[serde(skip_serializing_if="Option::is_none")]
    pub enddate: Option<String>,
}

pub type DnsOpts = HashMap<Dn, DirectOptions>;

#[derive(Serialize, PartialEq, Eq, Debug)]
pub struct SubjectAttrs {
    pub attrs: MonoAttrs,

    #[serde(skip_serializing_if="Option::is_none")]
    pub sgroup_id: Option<String>,

    #[serde(flatten)]
    pub options: DirectOptions,
}

pub type Subjects = BTreeMap<Dn, SubjectAttrs>;

#[derive(Serialize, PartialEq, Eq, Debug)]
pub struct SubjectsAndCount {
    pub count: usize,
    pub subjects: Subjects,
}

#[derive(Serialize, PartialEq, Eq, Debug)]
pub struct SgroupOutAndRight {
    pub attrs: MonoAttrs,

    pub sgroup_id: String,
    pub right: Option<Right>,
}


#[derive(Serialize, PartialEq, Eq, Debug)]
#[serde(rename_all = "camelCase")]
pub enum SgroupOutMore {
    Stem { children: SgroupsWithAttrs },
    Group { direct_members: Subjects },
    SynchronizedGroup { remote_sql_query: RemoteSqlQuery },
}

#[derive(Serialize, PartialEq, Eq, Debug)]
pub struct SgroupAndMoreOut {
    pub attrs: MonoAttrs,
    #[serde(flatten)]
    pub more: SgroupOutMore,

    pub parents: Vec<SgroupOutAndRight>,
    pub right: Right,
}

impl Mright {
    pub fn from_string(mright: &str) -> Result<Self, String> {
        serde::json::from_str(&format!(r#""{}""#, mright)).map_err(|_| format!("invalid mright {}", mright))
    }
    fn to_string(self) -> &'static str {
        match self {
            Self::Member => "member",
            Self::Reader => "reader",
            Self::Updater => "updater",
            Self::Admin => "admin",
        }
    }
    pub fn to_attr(self) -> String {
        format!("memberURL;x-{}", self.to_string())
    }
    pub fn to_attr_synchronized(self) -> String {
        format!("memberURL;x-{};x-remote", self.to_string())
    }
    pub fn list() -> Vec<Self> { vec![Self::Member, Self::Reader, Self::Updater, Self::Admin] }     
}
impl Right {
    pub fn from_string(right: &str) -> Result<Self, String> {
        serde::json::from_str(&format!(r#""{}""#, right)).map_err(|_| format!("invalid right {}", right))
    }
    // NB: best right first
    pub fn to_allowed_rights(self) -> Vec<Self> {
        match self {
            Self::Reader => vec![Self::Admin, Self::Updater, Self::Reader],
            Self::Updater => vec![Self::Admin, Self::Updater],
            Self::Admin => vec![Self::Admin],
        }
    }
    pub fn to_allowed_attrs(self) -> Vec<String> {
        self.to_allowed_rights().iter().map(|r| r.to_attr()).collect()        
    }
    pub fn to_mright(self) -> Mright {
        match self {
            Self::Reader => Mright::Reader,
            Self::Updater => Mright::Updater,
            Self::Admin => Mright::Admin,
        }
    }
    pub fn to_attr(self) -> String {
        self.to_mright().to_attr()
    }
}

#[derive(Debug)]
pub enum LoggedUser {
    TrustedAdmin,
    User(String),
}

impl LoggedUser {
    pub fn to_str(&self) -> &str {
        match self {
            LoggedUser::TrustedAdmin => "TrustedAdmin",
            LoggedUser::User(user) => user,
        }
    }
} 

#[derive(Debug)]
pub enum LoggedUserUrls {
    TrustedAdmin,
    User(HashSet<String>),
}

pub struct CfgAndLU<'a> {
    pub cfg: &'a Config,
    pub user: LoggedUser,
}

#[derive(Debug, Deserialize, Serialize, PartialEq, Eq, Clone)]
pub struct ToSubjectSource {
    pub ssdn: Dn,
    pub id_attr: String,
}

#[derive(Debug, Deserialize, Serialize, PartialEq, Eq, Clone)]
pub struct RemoteSqlQuery {
    pub remote_cfg_name: String, 
    pub select_query: String, // query returns either a DN or a string to transform into DN using ToSubjectSource
    pub to_subject_source: Option<ToSubjectSource>,
}

File: server/my_types.ts
import _ from 'lodash'
import { SessionOptions } from 'express-session';
import { Options as SessionFileStoreOptions } from 'session-file-store';
import { Periodicity } from './periodicity';


export type Option<T> = T | undefined

export const hOption = {
    isSome : <T>(v: Option<T>) => (
        v !== undefined && v !== null
    ),
    isGT : <T>(v1: Option<T>, v2: Option<T>) => (
        v1 !== undefined && v1 !== null && (
            v2 === undefined || v2 === null || v1 > v2
        )
    ),
}

export type MyMap<K extends string, V> = Partial<Record<K, V>>

export const hMyMap = {
    each: <K extends string, V>(map: MyMap<K, V>, cb: (v:V, k:K) => void) => {
        for (const k in map) cb(map[k] as V, k)
    },
    eachAsync: async <K extends string, V>(map: MyMap<K, V>, cb : (v:V, k:K) => Promise<void>) => {
        for (const k in map) await cb(map[k] as V, k)
    },
    mapToArray: <K extends string, V, R>(map: MyMap<K, V>, cb: (v:V, k:K) => R) => {
        const r: R[] = []
        for (const k in map) {
            r.push(cb(map[k] as V, k))
        }
        return r
    },
    mapValues: <K extends string, V, V_>(map: MyMap<K, V>, cb: (v:V, k:K) => V_) => (
        _.mapValues(map, cb) as MyMap<K,V_>
    ),
    filter: <K extends string, V>(map: MyMap<K, V>, cb: (v:V, k:K) => boolean) => (
        // @ts-expect-error
        _.pickBy(map, cb) as MyMap<K,V>
    ),
    /*filter_map: <K extends string, K_ extends string, V, V_>(map: MyMap<K, V>, cb: (v:V, k:K) => [K_,V_]) => {
        let r: MyMap<K_,V_> = {}
        for (const k in map) {
            cb(map[k], k)?.oMap(([k_, v_]) => r[k_] = v_)
        }
        return r
    },*/
    fromPairs: <K extends string, V>(pairs: [K,V][]) => {
        let r: MyMap<K,V> = {}
        for (const [k,v] of pairs) r[k] = v
        return r
    },
    // same as _.fromPairs(_.compact(...)), but more perfomant
    fromOptionPairs: <K extends string, V>(pairs: Option<[K,V]>[]) => {
        const r: MyMap<K,V> = {}
        for (const pair of pairs) {
            if (pair) r[pair[0]] = pair[1]
        }
        return r
    },
    firstEntry: <K extends string, V>(map: MyMap<K, V>) => (
        Object.entries(map)?.[0] as Option<[K,V]>
    ),
    firstValue: <K extends string, V>(map: MyMap<K, V>) => (
        Object.values(map)?.[0] as Option<V>
    ),
    compact: <K extends string, V>(map: MyMap<K,Option<V>>) => (
        _.pickBy(map, hOption.isSome) as MyMap<K,V>
    ),
    length: (map: MyMap<any, unknown>) => (
        _.size(map)
    ),
    keys: <K extends string, V>(map: MyMap<K, V>) => (
        Object.keys(map) as K[]
    ),
    values: <K extends string, V>(map: MyMap<K, V>) => (
        Object.values(map) as V[]
    ),
}


export type MySet<T> = T[]

export const default_separator = "."
export const default_root_id = ""

export const ldap_config_checker = (cfg : LdapConfig) => {   
    if (!sgroup_sscfg_raw(cfg)) {
        throw "expected ''ldap.groups_dn'' to be listed in ''ldap.subject_sources''"
    }
}

export interface StemConfig {
    filter: string,
    separator?: string,
    root_id?: string,
}

export interface SubjectSourceConfig {
    dn : FlavorDn,
    name : string,
    /** default template is "display first non empty `display_attrs` value" */
    vue_template ?: string,
    vue_template_if_ambiguous ?: string,    
    display_attrs : string[],
    id_attrs ?: string[],

    search_filter : string,
}

interface SgroupAttrTexts {
    label: string,
    description?: string,
    only_in_stem?: string,   
    vue_template?: string,    
    input_attrs?: MyMap<string, string>
}

export interface LdapConfig {
    connect: { uri: string[], dn: FlavorDn, password: string, verbose?: boolean },
    base_dn: FlavorDn,
    /** DN of the branch containing groups&stems groupald is working on. More info comes from the corresponding subject_source */
    groups_dn: FlavorDn,
    stem_object_classes: MySet<string>,
    group_object_classes: MySet<string>,
    sgroup_filter?: string, // needed if groupad does not own all groups in groups_dn
    group_filter: string,
    stem: StemConfig,
    subject_sources: SubjectSourceConfig[],
    groups_flattened_attr: MyMap<Mright, string>,
    sgroup_attrs: MyMap<string, SgroupAttrTexts>,
}
export interface LdapConfigOut {
    groups_dn: Dn,
    subject_sources: SubjectSourceConfig[],
    sgroup_attrs: MyMap<string, SgroupAttrTexts>,
}

const sgroup_sscfg_raw = (self: LdapConfig) => (
    self.subject_sources.find(sscfg => sscfg.dn === self.groups_dn)
)
export const hLdapConfig = {
    /** returns the ldap.subject_sources entry for the groups&stems groupald is working on */
    sgroup_sscfg: (self: LdapConfig) => {
        const sscfg = sgroup_sscfg_raw(self)
        if (!sscfg) {
            throw "internal error (should be checked as startup)"
        }
        return sscfg
    },
    /** export LdapConfig to the Vue.js UI */
    to_js_ui: (self: LdapConfig): LdapConfigOut => (
        { groups_dn: toDn(self.groups_dn), subject_sources: self.subject_sources, sgroup_attrs: self.sgroup_attrs }
    ),
}

/** known remote drivers */
export const remoteSqlDrivers = [ 'mysql', 'oracle', 'postgresql' ] as const
export type RemoteSqlDriver = (typeof remoteSqlDrivers)[number]

export interface RemoteSqlConfig {
    driver: RemoteSqlDriver,

    host: string,
    port?: number, // u16
    user: string,
    password: string,
    
    db_name: string,
}
export interface RemoteLdapConfig {
    driver: 'ldap'
    connect: LdapConfig['connect']
    search_branch?: FlavorDn
}
export type RemoteConfig = {
    periodicity: Periodicity
} & (RemoteLdapConfig | RemoteSqlConfig)

/** helpers to work on RemoteConfig */
export const hRemoteConfig = {
    export: (self: RemoteConfig) => (
        _.omit(self, ['user', 'password', 'connect.user', 'connect.password'])
    ),
}

export interface Config {
    trusted_auth_bearer?: string,
    log_dir?: string,
    cas: { prefix_url: string },
    trust_proxy?: string,
    session_store: {
        options: SessionOptions,
        file_store: SessionFileStoreOptions,
    },
    //#[serde(deserialize_with = "ldap_config_checker")] 
    ldap: LdapConfig,
    remotes: MyMap<string, RemoteConfig>,
    additional_periodicities: Periodicity[]
    remote_forced_periodicity_attr: string
}

/** member or right */
export type Mright = 'member' | Right
export type Right = 'reader' | 'updater' | 'admin'

export const hMright = {
    to_attr: (self: Mright) => (
        `memberURL;x-${self}`
    ),
    attr_synchronized: `memberURL;x-member-remote`,
    list: (): Mright[] => (
        [ 'member', 'reader', 'updater', 'admin' ]
    ),
}

/** NB: best rights first */
const to_allowed_rights = (self: Right): Right[] => {
    switch (self) {
        case 'reader': return ['admin', 'updater', 'reader']
        case 'updater': return ['admin', 'updater']
        case 'admin': return ['admin']
    }
}

/** helpers to work on rights */
export const hRight = {
    to_allowed_rights,
    // NB: best right first
    to_allowed_attrs: (self: Right): string[] => (
        to_allowed_rights(self).map(hMright.to_attr)
    ),
    to_attr: hMright.to_attr,
    max: (a: Option<Right>, b: Option<Right>) => {
        if (a === 'admin' || b === 'admin') return 'admin'
        if (a === 'updater' || b === 'updater') return 'updater'
        if (a === 'reader' || b === 'reader') return 'reader'
        return undefined
    },
    max_: (l: Right[]) => {
        let r: Option<Right>
        for (const a of l) {
            r = hRight.max(r, a)
        }
        return r
    },
    list: (): Right[] => (
        [ 'reader', 'updater', 'admin' ]
    ),
}

// same as ldapjs.Change "operation"
export type MyMod = 'add' | 'delete' | 'replace'

// https://stackoverflow.com/questions/56737033/how-to-define-an-opaque-type-in-typescript
export type FlavorDn = string & { _type?: "Dn" }
export type Dn = string & { _type: "Dn" } // fake field to enforce types: you can't convert to "Dn" without casting
export const toDns = (dns: string[]) => dns as Dn[]
export const toDn = (dn: string) => dn as Dn

/** members or rights to add/remove/replace (with optional member options) */
export type MyMods = MyMap<Mright, MyMap<MyMod, DnsOpts>>

export type MonoAttrs = MyMap<string, string>;
export type MultiAttrs = MyMap<string, string[]>;

export type SgroupsWithAttrs = MyMap<string, MonoAttrs>;

export interface DirectOptions {
    enddate?: string,
}

export type DnsOpts = MyMap<Dn, DirectOptions>;

export interface SubjectAttrs {
    attrs: MonoAttrs,
    sgroup_id?: string,
    options: DirectOptions,
}

export type Subjects = MyMap<Dn, SubjectAttrs>;

export type SubjectsOrNull = MyMap<Dn, SubjectAttrs | null>;

export interface SubjectsAndCount {
    count: number, // interger
    subjects: SubjectsOrNull,
}

/** group/stem id & attributes + loggedUser right on this group/stem */
export interface SgroupOutAndRight {
    attrs: MonoAttrs,

    sgroup_id: string,
    right?: Right,
}

/** stem children or group direct_members or sync group definition */
export type SgroupOutMore = 
    { stem: { children: SgroupsWithAttrs } } |
    { group: { direct_members: SubjectsOrNull } } |
    { synchronizedGroup: { remote_query: RemoteQuery, last_sync_date: Option<Date> } }

/** group/stem attributes + parents + loggedUser right on this group/stem + stem children or group direct_members or sync group definition */
export type SgroupAndMoreOut = SgroupOutMore & {
    attrs: MonoAttrs,
    parents: SgroupOutAndRight[],
    right: Right,
}

export type LoggedUser = { TrustedAdmin: true } | { User: string }
/** LoggerUser helpers */
export const hLoggedUser = {
    toString: (self: LoggedUser) => (
        'User' in self ? self.User : "TrustedAdmin"
    ),
}
export type LoggedUserDn = { TrustedAdmin: true } | { User: Dn }

/** to transform values returned by SELECT query into DN */
export interface ToSubjectSource {
    /** branch DN to search (eg: ou=people,...). If must be listed in conf.ldap.subject_sources */
    ssdn: Dn,
    /** attribute to use to find values. If not given, all sscfg.id_attrs will be searched */
    id_attr?: string,
}

export interface RemoteLdapQuery {
    /** empty string means the LDAP server used for groups&subjects */
    remote_cfg_name: string
    forced_periodicity?: Periodicity
    /** LDAP branch to search (defaults to conf.ldap.base_dn if remote_cfg_name is "") */
    DN?: string
    /** LDAP filter on the branch */
    filter?: string
    /** usually unset (since the DNs matching the branch & filter are directly used). Useful examples: "member", "seeAlso"... */
    attribute?: string
}
export interface RemoteSqlQuery {
    remote_cfg_name: string, 
    forced_periodicity?: Periodicity
    /** query which returns either a DN or a string to transform into DN using ToSubjectSource */
    select_query: string, 
    /** how to transform values into a DN */
    to_subject_source?: ToSubjectSource,
}
export type RemoteQuery = RemoteLdapQuery | RemoteSqlQuery
export const isRqSql = (rq: Option<RemoteQuery>): rq is RemoteSqlQuery => (
    rq ? "select_query" in rq : false
)
export const toRqSql = (rq: Option<RemoteQuery>) => (
    isRqSql(rq) ? rq : undefined
)

export interface TestRemoteQuery {
    count: number,
    values: string[],
    values_truncated: boolean,
    ss_guess: Option<[ToSubjectSource, Subjects]>,
}

export type SgroupLog = { who: string, when: Date, action: string } & Record<string, any>

export interface SgroupLogs { 
    last_log_date: Option<Date>
    whole_file: boolean
    logs: SgroupLog[]
}
